### ISSUE TITLE
parsing python like languages

### ISSUE THREAD

**Issue by schneidersoft** [arithy/packcc#79]
How would I parse python like languages where indentation is used to handle scope?

**Comment by dolik-rce** [comment]
Hello @schneidersoft,

the indentation is just another character for the parser, so there is not much different to other parsers. You can also use references to make sure the indentation on each line is correct (at least in simple cases, more complex code would probably need to check this in code).

Here is a very simple grammar to parse python-like function definition:
```
function <- "def " identifier "():\n" <indent> statement "\n" ($1 statement "\n")*

identifier <- [_a-zA-Z][_a-zA-Z0-9]*

indent <- " "+ / "\t"+

statement <- [^ \t\n][^\n]*
```

It would report syntax error if you run it on incorrectly indented code.

**Comment by dolik-rce** [comment]
PS: Here is a full python grammar (just in slightly different format than the one used by PackCC), if you need some inspiration: https://docs.python.org/3/reference/grammar.html

**Comment by schneidersoft** [comment]
Right. I was wondering how to integrate a tokenizer that would produce the INDENT end DEDENT tokens packcc would then be able to use.

**Comment by mingodad** [comment]
Look here https://github.com/exaloop/codon/blob/develop/codon/parser/peg/grammar.peg how it uses https://github.com/yhirose/cpp-peglib to parse a Python like language.

**Comment by arithy** [comment]
Hello @mingodad , I have introduced 'programmable predicates' and it should be fulfill your needs. It can be realized by restricting to access only 'marker variables', which are guaranteed not to cause parsing context corruption. If you still have interests, please try it.
